File: shellsight/probes/dotnetmem/src/Contract.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace ShellSight.DotnetMem;

// The input half of the probe contract (mirrors Go finding.TargetSpec).
public sealed class TargetSpec
{
    [JsonPropertyName("host")] public string Host { get; set; } = "";
    [JsonPropertyName("pids")] public int[]? Pids { get; set; }
    [JsonPropertyName("webroots")] public string[]? Webroots { get; set; }
    [JsonPropertyName("dump_path")] public string DumpPath { get; set; } = "";
}

// The output half (mirrors Go finding.Finding). JSON property names == Go json tags.
public sealed class Finding
{
    [JsonPropertyName("schema_version")] public string SchemaVersion { get; set; } = "1.0";
    [JsonPropertyName("id")] public string Id { get; set; } = "";
    [JsonPropertyName("host")] public string Host { get; set; } = "";
    [JsonPropertyName("view")] public string View { get; set; } = "dotnet-mem";
    [JsonPropertyName("target")] public Target Target { get; set; } = new();
    [JsonPropertyName("artifact")] public Artifact Artifact { get; set; } = new();
    [JsonPropertyName("detection")] public Detection Detection { get; set; } = new();
    [JsonPropertyName("score")] public int Score { get; set; }
    [JsonPropertyName("tier")] public string Tier { get; set; } = "suspicious";
    [JsonPropertyName("classification")] public Classification Classification { get; set; } = new();
    [JsonPropertyName("artifacts")] public Artifacts Artifacts { get; set; } = new();
    [JsonPropertyName("context")] public Dictionary<string, string>? Context { get; set; }
}

public sealed class Target
{
    [JsonPropertyName("kind")] public string Kind { get; set; } = "process";
    [JsonPropertyName("process")] public ProcessRef? Process { get; set; }
    [JsonPropertyName("file")] public FileRef? File { get; set; }
}
public sealed class ProcessRef
{
    [JsonPropertyName("pid")] public int Pid { get; set; }
    [JsonPropertyName("name")] public string Name { get; set; } = "";
    [JsonPropertyName("app_pool")] [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)] public string AppPool { get; set; } = "";
    [JsonPropertyName("bitness")] [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)] public string Bitness { get; set; } = "";
}
public sealed class FileRef
{
    [JsonPropertyName("path")] public string Path { get; set; } = "";
    [JsonPropertyName("sha256")] public string Sha256 { get; set; } = "";
}
public sealed class Artifact
{
    [JsonPropertyName("kind")] public string Kind { get; set; } = "";
    [JsonPropertyName("identity")] public string Identity { get; set; } = "";
    [JsonPropertyName("location")] public string Location { get; set; } = "";
}
public sealed class Detection
{
    [JsonPropertyName("basis")] public string Basis { get; set; } = "structural-heuristic";
    [JsonPropertyName("knowledge_ref")] public string KnowledgeRef { get; set; } = "";
    [JsonPropertyName("evidence")] public string Evidence { get; set; } = "";
    [JsonPropertyName("allowlisted")] public bool Allowlisted { get; set; }
}
// The architected-in hook: always emitted, always null in v1 (stable schema for the future classifier).
public sealed class Classification
{
    [JsonPropertyName("family")] public string? Family { get; set; }
    [JsonPropertyName("capability")] public string[]? Capability { get; set; }
    [JsonPropertyName("source")] public string? Source { get; set; }
    [JsonPropertyName("confidence")] public double Confidence { get; set; }
}
public sealed class Artifacts
{
    [JsonPropertyName("raw")] [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)] public string Raw { get; set; } = "";
    [JsonPropertyName("decompiled")] [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)] public string Decompiled { get; set; } = "";
    [JsonPropertyName("features")] [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)] public string Features { get; set; } = "";
}

public static class Contract
{
    private static readonly JsonSerializerOptions Opts = new()
    {
        // Property names are set explicitly via [JsonPropertyName]; do not auto-rename.
        DefaultIgnoreCondition = JsonIgnoreCondition.Never, // classification nulls MUST be emitted
        WriteIndented = false,
    };
    private static readonly JsonSerializerOptions ParseOpts = new() { PropertyNameCaseInsensitive = true };

    public static string Serialize(IEnumerable<Finding> findings) => JsonSerializer.Serialize(findings, Opts);
    public static TargetSpec ParseSpec(string json) =>
        string.IsNullOrWhiteSpace(json) ? new TargetSpec() : (JsonSerializer.Deserialize<TargetSpec>(json, ParseOpts) ?? new TargetSpec());
}

File: shellsight/probes/javamem/src/com/shellsight/javamem/Contract.java
package com.shellsight.javamem;

import com.google.gson.*;
import com.google.gson.annotations.SerializedName;
import java.util.List;
import java.util.Map;

class TargetSpec {
    @SerializedName("host") String host = "";
    @SerializedName("pids") int[] pids;
    @SerializedName("webroots") String[] webroots;
    @SerializedName("dump_path") String dumpPath = "";
}
class Finding {
    @SerializedName("schema_version") String schemaVersion = "1.0";
    @SerializedName("id") String id = "";
    @SerializedName("host") String host = "";
    @SerializedName("view") String view = "java-mem";
    @SerializedName("target") Target target = new Target();
    @SerializedName("artifact") Artifact artifact = new Artifact();
    @SerializedName("detection") Detection detection = new Detection();
    @SerializedName("score") int score;
    @SerializedName("tier") String tier = "suspicious";
    @SerializedName("classification") Classification classification = new Classification();
    @SerializedName("artifacts") Artifacts artifacts = new Artifacts();
    @SerializedName("context") Map<String,String> context;
}
class Target { @SerializedName("kind") String kind = "process"; @SerializedName("process") ProcessRef process; @SerializedName("file") FileRef file; }
class ProcessRef { @SerializedName("pid") int pid; @SerializedName("name") String name = ""; }
class FileRef { @SerializedName("path") String path = ""; @SerializedName("sha256") String sha256 = ""; }
class Artifact { @SerializedName("kind") String kind = ""; @SerializedName("identity") String identity = ""; @SerializedName("location") String location = ""; }
class Detection { @SerializedName("basis") String basis = "structural-heuristic"; @SerializedName("knowledge_ref") String knowledgeRef = ""; @SerializedName("evidence") String evidence = ""; @SerializedName("allowlisted") boolean allowlisted; }
class Classification { @SerializedName("family") String family; @SerializedName("capability") String[] capability; @SerializedName("source") String source; @SerializedName("confidence") double confidence; }
class Artifacts { @SerializedName("raw") String raw = ""; @SerializedName("decompiled") String decompiled = ""; @SerializedName("features") String features = ""; }

final class Contract {
    // serializeNulls so classification.family/source/capability emit as JSON null (Go requires the key
    // present — the architected classifier hook). Go's encoding/json maps null -> zero value, safe to parse.
    private static final Gson GSON = new GsonBuilder().serializeNulls().disableHtmlEscaping().create();
    static String serialize(List<Finding> findings) { return GSON.toJson(findings); }

    /**
     * Probe coverage, mirroring Go's finding.ProbeCoverage. Field names are the JSON contract, so
     * they are spelled the way the core reads them.
     */
    static final class ProbeCoverage {
        String status;                 // ran | degraded | failed | n/a
        String reason;
        int targets_scanned;
        /**
         * The unbounded-gap flag. The core turns this into `incomplete`, which turns a clean tier
         * into `unknown` and exit 5, because a sweep that stopped early cannot say what was in the
         * part it never reached.
         *
         * <p>A primitive, so it always emits explicitly. GSON here is built with serializeNulls()
         * for the classification hook, which applies to the whole tree -- a boxed Boolean left null
         * emits as `truncated: null`, not as an absent key, so boxing bought nothing but ambiguity.
         * Go reads either into false; an explicit false is what a reader can act on.
         */
        boolean truncated;
    }

    /** The object form of the probe contract: {findings, coverage}. */
    static final class ProbeOutput {
        List<Finding> findings;
        ProbeCoverage coverage;
    }

    /**
     * Serialize the OBJECT form.
     *
     * <p>The probe used to print a bare findings array -- the legacy form the core still accepts --
     * which had no place to put a coverage statement, so budget exhaustion could not be reported
     * however much the probe knew about it. GSON is configured with serializeNulls for the
     * classification hook, so `truncated` is dropped explicitly rather than emitted as null.
     */
    static String serializeOutput(List<Finding> findings, ProbeCoverage coverage) {
        ProbeOutput out = new ProbeOutput();
        out.findings = findings;
        out.coverage = coverage;
        return GSON.toJson(out);
    }
    static TargetSpec parseSpec(String json) {
        if (json == null || json.isBlank()) return new TargetSpec();
        TargetSpec s = GSON.fromJson(json, TargetSpec.class);
        return s != null ? s : new TargetSpec();
    }
}
